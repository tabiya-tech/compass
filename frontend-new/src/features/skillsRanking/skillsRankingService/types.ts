export type SkillsRankingMetrics = {
  cancelled_after?: string,
  succeeded_after?:string,
  puzzles_solved?:number,
  correct_rotations?:number,
  clicks_count?:number,
}