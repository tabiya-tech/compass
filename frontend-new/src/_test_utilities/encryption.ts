import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
