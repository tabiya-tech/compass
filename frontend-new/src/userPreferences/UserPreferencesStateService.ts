import { SensitivePersonalDataRequirement, UserPreference, Language } from "src/userPreferences/UserPreferencesService/userPreferences.types";
import {
  QUESTION_KEYS,
} from "src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service.types";
import * as Sentry from "@sentry/react";

export default class UserPreferencesStateService {
  private static instance: UserPreferencesStateService;
  private userPreferences: UserPreference | null = null;

  private constructor() {
  }

  public static getInstance(): UserPreferencesStateService {
    if (!UserPreferencesStateService.instance) {
      UserPreferencesStateService.instance = new UserPreferencesStateService();
    }
    return UserPreferencesStateService.instance;
  }

  /**
   * Retrieves the current user preferences.
   * It returns a deep copy of the user preferences object to prevent direct modification.
   * @returns {UserPreference | null} The current user preferences object or null if no user preferences are set or
   * an error occurs while cloning the object.
   */
  public getUserPreferences(): UserPreference | null {
    try {
      return this.cloneUserPreferences(this.userPreferences);
    } catch (e) {
      console.error(new Error("Error in getUserPreferences", { cause: e }));
      return null;
    }
  }

  /**
   * Sets the current user preferences.
   * It stores a deep copy of the user preferences object to prevent direct modification.
   * @param preferences
   */
  public setUserPreferences(preferences: UserPreference): void {
    // Store a deep copy of the user preferences object to prevent direct modification
    this.userPreferences = this.cloneUserPreferences(preferences);
    try {
      // Set the session context for Sentry so that we can track the session id in errors
      Sentry.setContext("session", {
        session_id: this.getActiveSessionId() ?? "UNKNOWN"
      });
    } catch (err) {
      console.error(new Error(`Failed to set sentry context`, { cause: err }));
    }
  }

  public clearUserPreferences(): void {
    this.userPreferences = null;
  }

  public getActiveSessionId(): number | null {
    return this.userPreferences?.sessions.length ? this.userPreferences?.sessions[0] : null;
  }

  public activeSessionHasOverallFeedback(): boolean {
    const activeSessionId = this.getActiveSessionId();
    if (activeSessionId === null) {
      return false;
    }
    const answered_questions = this.userPreferences!.user_feedback_answered_questions
    if (Object.keys(answered_questions).length === 0) {
      return false;
    }

    if (!answered_questions[activeSessionId]) {
      return false;
    }

    return answered_questions[activeSessionId].some(
      (question) => question !== QUESTION_KEYS.CUSTOMER_SATISFACTION
    );
  }
  
  public activeSessionHasCustomerSatisfactionRating(): boolean {
    const activeSessionId = this.getActiveSessionId();

    if (activeSessionId === null) {
      return false;
    }

    // userPreferences.user_feedback_answered_questions is an object with session ids as keys,
    // and it is never undefined, but it can be an empty object: {}.
    // For Postel's Law: assert that the object is defined.
    const answered_questions = this.userPreferences!.user_feedback_answered_questions
    if (!answered_questions || Object.keys(answered_questions).length === 0) {
      return false;
    }

    if (!answered_questions[activeSessionId]) {
      return false;
    }

    return answered_questions[activeSessionId].includes(QUESTION_KEYS.CUSTOMER_SATISFACTION)
  }

  private cloneUserPreferences(preferences: UserPreference | null): UserPreference | null {
    if (!preferences) {
      return preferences;
    }

    const strObj = JSON.stringify(preferences);
    return JSON.parse(strObj, (key, value) => {
      if (key === "accepted_tc") {
        return new Date(value);
      }
      if (key === "sensitive_personal_data_requirement") {
        return SensitivePersonalDataRequirement[value as keyof typeof SensitivePersonalDataRequirement];
      }
      if (key === "language") {
        return Language[value as keyof typeof Language];
      }
      return value;
    });
  }
}
