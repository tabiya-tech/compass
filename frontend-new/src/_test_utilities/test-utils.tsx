// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
