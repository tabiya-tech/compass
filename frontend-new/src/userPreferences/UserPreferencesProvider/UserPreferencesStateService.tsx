import { UserPreference } from "src/userPreferences/UserPreferencesService/userPreferences.types";
import { userPreferencesService } from "src/userPreferences/UserPreferencesService/userPreferences.service";

class UserPreferencesService {
  private static instance: UserPreferencesService;
  private userPreferences: UserPreference | null = null;

  private constructor() {}

  public static getInstance(): UserPreferencesService {
    if (!UserPreferencesService.instance) {
      UserPreferencesService.instance = new UserPreferencesService();
    }
    return UserPreferencesService.instance;
  }

  public async loadPreferences(user_id: string) {
    try {
      if (!this.userPreferences?.accepted_tc) {
        this.userPreferences = await userPreferencesService.getUserPreferences(user_id);
      }
    } catch (error) {
      console.error("Error loading user preferences", error);
    }
  }

  public getUserPreferences(): UserPreference | null {
    return this.userPreferences;
  }

  public setUserPreferences(preferences: UserPreference): void {
    this.userPreferences = preferences;
  }

  public clearUserPreferences(): void {
    this.userPreferences = null;
  }
}

export const userPreferencesStateService = UserPreferencesService.getInstance();
