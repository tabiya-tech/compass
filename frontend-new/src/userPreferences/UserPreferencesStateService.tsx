import { UserPreference } from "src/userPreferences/UserPreferencesService/userPreferences.types";

export default class UserPreferencesStateService {
  private static instance: UserPreferencesStateService;
  private userPreferences: UserPreference | null = null;

  private constructor() {}

  public static getInstance(): UserPreferencesStateService {
    if (!UserPreferencesStateService.instance) {
      UserPreferencesStateService.instance = new UserPreferencesStateService();
    }
    return UserPreferencesStateService.instance;
  }

  public getUserPreferences(): UserPreference | null {
    return this.userPreferences;
  }

  public setUserPreferences(preferences: UserPreference): void {
    this.userPreferences = preferences;
  }

  public clearUserPreferences(): void {
    this.userPreferences = null;
  }

  public getActiveSessionId(): number | null {
    return this.userPreferences?.sessions.length ? this.userPreferences?.sessions[0] : null;
  }

  public activeSessionHasFeedback(): boolean {
    const activeSessionId = this.getActiveSessionId();
    if (activeSessionId === null) {
      return false;
    }
    return this.userPreferences!.sessions_with_feedback.includes(activeSessionId);
  }
}
