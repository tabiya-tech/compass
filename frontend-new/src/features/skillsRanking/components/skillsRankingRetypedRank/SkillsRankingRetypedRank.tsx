import React, { useCallback, useContext, useEffect, useRef, useState } from "react";
import { Box, useTheme } from "@mui/material";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";
import ChatBubble from "src/chat/chatMessage/components/chatBubble/ChatBubble";
import { MessageContainer } from "src/chat/chatMessage/compassChatMessage/CompassChatMessage";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import {
  SkillsRankingPhase,
  SkillsRankingState,
  getLatestPhaseName,
} from "src/features/skillsRanking/types";
import { SkillsRankingService } from "src/features/skillsRanking/skillsRankingService/skillsRankingService";
import { SkillsRankingError } from "src/features/skillsRanking/errors";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { IsOnlineContext } from "src/app/isOnlineProvider/IsOnlineProvider";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import TypingChatMessage from "src/chat/chatMessage/typingChatMessage/TypingChatMessage";
import { AnimatePresence, motion } from "framer-motion";
import { useAutoScrollOnChange } from "src/features/skillsRanking/hooks/useAutoScrollOnChange";
import ChatMessageFooterLayout from "src/chat/chatMessage/components/chatMessageFooter/ChatMessageFooterLayout";
import Timestamp from "src/chat/chatMessage/components/chatMessageFooter/components/timestamp/Timestamp";
import { getJobPlatformUrl } from "src/features/skillsRanking/constants";
import SkillsRankingSlider from "src/features/skillsRanking/components/skillsRankingSlider/SkillsRankingSlider";
import { shouldSkipMarketDisclosure } from "src/features/skillsRanking/utils/createMessages";

const uniqueId = "eb90de4c-2462-4b6d-8b9c-1b5c6ae64129";
const TYPING_DURATION_MS = 5000;

export const DATA_TEST_ID = {
  SKILLS_RANKING_RETYPED_RANK_CONTAINER: `skills-ranking-retyped-rank-container-${uniqueId}`,
  SKILLS_RANKING_RETYPED_RANK_SLIDER: `skills-ranking-retyped-rank-slider-${uniqueId}`,
  SKILLS_RANKING_RETYPED_RANK_SUBMIT_BUTTON: `skills-ranking-retyped-rank-submit-button-${uniqueId}`,
};

export interface SkillsRankingRetypedRankProps {
  onFinish: (skillsRankingState: SkillsRankingState) => Promise<void>;
  skillsRankingState: SkillsRankingState;
}

export const SKILLS_RANKING_RETYPED_RANK_MESSAGE_ID = `skills-ranking-retyped-rank-message-${uniqueId}`;

const SkillsRankingRetypedRank: React.FC<Readonly<SkillsRankingRetypedRankProps>> = ({
  onFinish,
  skillsRankingState,
}) => {
  const theme = useTheme();
  const [value, setValue] = useState(0);
  const [startedEditing, setStartedEditing] = useState(false);
  const [submitted, setSubmitted] = useState(false);
  const [showTyping, setShowTyping] = useState(false);
  const hasFinishedRef = useRef(false);

  const activeSessionId = UserPreferencesStateService.getInstance().getActiveSessionId();
  const isOnline = useContext(IsOnlineContext);
  const { enqueueSnackbar } = useSnackbar();

  const currentPhase = getLatestPhaseName(skillsRankingState);
  const scrollRef = useAutoScrollOnChange(showTyping);

  // Use the utility function instead of local variable
  const shouldNotShow = shouldSkipMarketDisclosure(skillsRankingState.experiment_group);

  const handleUpdateState = useCallback(async () => {
    if (!activeSessionId) {
      console.error(new SkillsRankingError("Active session ID is not available."));
      return;
    }
    try {
      const newState = await SkillsRankingService.getInstance().updateSkillsRankingState(
        activeSessionId,
        SkillsRankingPhase.COMPLETED,
        undefined,
        value
      );
      setSubmitted(true);
      setShowTyping(true);
      setTimeout(() => setShowTyping(false), TYPING_DURATION_MS);
      setTimeout(() => {
        if (!hasFinishedRef.current) {
          hasFinishedRef.current = true;
          onFinish(newState);
        }
      }, TYPING_DURATION_MS + 300);
    } catch (err) {
      console.error("Failed to update state:", err);
      enqueueSnackbar("Something went wrong. Please try again.", { variant: "error" });
    }
  }, [activeSessionId, value, onFinish, enqueueSnackbar]);

  const handleSubmit = async () => {
    if (!submitted && currentPhase === SkillsRankingPhase.RETYPED_RANK) {
      await handleUpdateState();
    }
  };

  useEffect(() => {
    if (currentPhase !== SkillsRankingPhase.RETYPED_RANK) {
      setValue(skillsRankingState.retyped_rank_percentile ?? 0);
    }
  }, [skillsRankingState, currentPhase]);

  useEffect(() => {
    if (shouldNotShow && !hasFinishedRef.current) {
      setShowTyping(true); // Show typing animation
      const typingTimer = setTimeout(async () => {
        if (!activeSessionId) {
          console.error(new SkillsRankingError("Active session ID is not available."));
          return;
        }

        if (currentPhase !== SkillsRankingPhase.RETYPED_RANK) {
          console.warn(`[RetypedRank] Skipping auto-submit. Phase is already ${currentPhase}`);
          return;
        }

        hasFinishedRef.current = true;
        try {
          const newState = await SkillsRankingService.getInstance().updateSkillsRankingState(
            activeSessionId,
            SkillsRankingPhase.COMPLETED,
            undefined,
            value
          );
          await onFinish(newState);
        } catch (err) {
          console.error("Failed to update state:", err);
          enqueueSnackbar("Something went wrong. Please try again.", { variant: "error" });
        }
      }, TYPING_DURATION_MS);

      return () => clearTimeout(typingTimer);
    }
  }, [shouldNotShow, activeSessionId, currentPhase, value, onFinish, enqueueSnackbar]);

  if (shouldNotShow) return <></>;

  return (
    <MessageContainer
      origin={ConversationMessageSender.COMPASS}
      ref={scrollRef}
      data-testid={DATA_TEST_ID.SKILLS_RANKING_RETYPED_RANK_CONTAINER}
      gap={theme.fixedSpacing(theme.tabiyaSpacing.md)}
    >
      {!shouldNotShow && (
        <Box sx={{ width: "100%" }}>
          <ChatBubble
            sender={ConversationMessageSender.COMPASS}
            message={
              <>
                In any case, if we do not think about other job seekers but again focus on those opportunities available
                to you, let's move to creating your skills profile that you can share with those employers in the next
                step.
                {"\n\n"}
                As a last question, let's remind ourselves of what I told you further above: check again what I said
                three messages ago, how many percent of opportunities on <strong>{getJobPlatformUrl()}</strong> do you
                fulfill the required & most relevant skills of?
              </>
            }
          >
            <Box padding={theme.fixedSpacing(theme.tabiyaSpacing.md)}>
              <SkillsRankingSlider
                value={value}
                onChange={(_, newVal) => {
                  setStartedEditing(true);
                  setValue(newVal as number);
                }}
                disabled={submitted || !isOnline || currentPhase !== SkillsRankingPhase.RETYPED_RANK}
                data-testid={DATA_TEST_ID.SKILLS_RANKING_RETYPED_RANK_SLIDER}
                aria-label="Retyped rank percentile slider"
              />

              <Box mt={theme.spacing(2)} textAlign="right">
                <PrimaryButton
                  onClick={handleSubmit}
                  disabled={
                    submitted || !startedEditing || !isOnline || currentPhase !== SkillsRankingPhase.RETYPED_RANK
                  }
                  data-testid={DATA_TEST_ID.SKILLS_RANKING_RETYPED_RANK_SUBMIT_BUTTON}
                >
                  Submit
                </PrimaryButton>
              </Box>
            </Box>
          </ChatBubble>

          <ChatMessageFooterLayout sender={ConversationMessageSender.COMPASS}>
            <Timestamp
              sentAt={
                skillsRankingState.phases[skillsRankingState.phases.length - 1]?.time || skillsRankingState.started_at
              }
            />
          </ChatMessageFooterLayout>
        </Box>
      )}

      <AnimatePresence mode="wait">
        {showTyping && (
          <motion.div
            key="typing"
            initial={{ opacity: 0, height: 0 }}
            animate={{ opacity: 1, height: "auto" }}
            exit={{ opacity: 0, height: 0 }}
            transition={{ duration: 0.3 }}
          >
            <TypingChatMessage />
          </motion.div>
        )}
      </AnimatePresence>
    </MessageContainer>
  );
};

export default SkillsRankingRetypedRank;
