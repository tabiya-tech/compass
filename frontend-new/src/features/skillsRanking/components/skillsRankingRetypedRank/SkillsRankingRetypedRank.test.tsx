// mute the console
import "src/_test_utilities/consoleMock";

import React from "react";
import { act, render, screen, fireEvent, waitFor } from "src/_test_utilities/test-utils";
import SkillsRankingRetypedRank, { DATA_TEST_ID } from "./SkillsRankingRetypedRank";
import {
  SkillsRankingExperimentGroups,
  SkillsRankingPhase,
  SkillsRankingState,
} from "src/features/skillsRanking/types";
import { SkillsRankingService } from "src/features/skillsRanking/skillsRankingService/skillsRankingService";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";

jest.mock("framer-motion", () => ({
  AnimatePresence: ({ children }: any) => <>{children}</>,
  motion: { div: ({ children, ...props }: any) => <div {...props}>{children}</div> },
}));

// Mock the slider component to a simple range input that forwards onChange(event, value)
jest.mock("src/features/skillsRanking/components/skillsRankingSlider/SkillsRankingSlider", () => ({
  __esModule: true,
  default: ({ onChange, value, ...props }: any) => (
    <input
      type="range"
      data-testid={props["data-testid"]}
      aria-label={props["aria-label"]}
      value={value}
      onChange={(e: any) => onChange(e, Number(e.target.value))}
      onInput={(e: any) => onChange(e, Number(e.target.value))}
    />
  ),
}));

jest.mock("src/features/skillsRanking/utils/createMessages", () => ({
  shouldSkipMarketDisclosure: jest.fn(() => false),
}));

describe("SkillsRankingRetypedRank", () => {
  beforeEach(() => {
    jest.useFakeTimers();
    Object.defineProperty(navigator, "onLine", { configurable: true, value: true });
    Object.defineProperty(HTMLElement.prototype, "scrollIntoView", {
      configurable: true,
      value: jest.fn(),
    });
    // Ensure disclosure is not skipped by default
    const { shouldSkipMarketDisclosure } = jest.requireMock("src/features/skillsRanking/utils/createMessages");
    shouldSkipMarketDisclosure.mockReturnValue(false);
  });
  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllMocks();
  });

  const flush = async () => {
    await act(async () => Promise.resolve());
  };

  const createState = (phase: SkillsRankingPhase, retyped?: number): SkillsRankingState => ({
    session_id: 1,
    experiment_group: SkillsRankingExperimentGroups.GROUP_2,
    phases: [{ name: phase, time: new Date().toISOString() }],
    score: {
      jobs_matching_rank: 0,
      comparison_rank: 0,
      comparison_label: "MIDDLE",
      calculated_at: new Date().toISOString(),
    },
    started_at: new Date().toISOString(),
    retyped_rank_percentile: retyped,
  });

  const setSliderValue = (value: number) => {
    const input = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_RETYPED_RANK_SLIDER) as HTMLInputElement;
    input.value = String(value);
    fireEvent.input(input, { target: { value: String(value) } });
    fireEvent.change(input, { target: { value: String(value) } });
  };

  test("should submit retyped rank and call service then onFinish (non-replay)", async () => {
    // GIVEN service mocks and a session id
    const givenSessionId = 55555;
    const mockUpdate = jest.fn().mockResolvedValue({} as SkillsRankingState);
    jest.spyOn(SkillsRankingService, "getInstance").mockReturnValue({
      updateSkillsRankingState: mockUpdate,
      getConfig: () => ({
        config: {
          compensationAmount: "$1",
          jobPlatformUrl: "x",
          shortTypingDurationMs: 1,
          defaultTypingDurationMs: 1,
          longTypingDurationMs: 1,
        },
      }),
    } as unknown as SkillsRankingService);
    jest
      .spyOn(UserPreferencesStateService, "getInstance")
      .mockReturnValue({ getActiveSessionId: () => givenSessionId } as any);
    const actualOnFinish = jest.fn().mockResolvedValue(undefined);
    const givenState = createState(SkillsRankingPhase.RETYPED_RANK);

    // WHEN render and change slider and submit
    render(<SkillsRankingRetypedRank onFinish={actualOnFinish} skillsRankingState={givenState} />);
    setSliderValue(55);
    await flush();
    const givenSubmitButton = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_RETYPED_RANK_SUBMIT_BUTTON);
    await waitFor(() => expect(givenSubmitButton).not.toBeDisabled());
    act(() => {
      givenSubmitButton.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    // THEN expect service and onFinish after typing delays
    act(() => {
      jest.runOnlyPendingTimers();
    }); // hide typing
    await flush();
    act(() => {
      jest.runOnlyPendingTimers();
    }); // call onFinish
    await flush();
    expect(mockUpdate).toHaveBeenCalledWith(givenSessionId, SkillsRankingPhase.COMPLETED, undefined, 55);
    expect(actualOnFinish).toHaveBeenCalled();
    expect(console.error).not.toHaveBeenCalled();
  });

  test("should auto-complete when market disclosure is skipped (effect path)", async () => {
    // GIVEN disclosure skipped
    const { shouldSkipMarketDisclosure } = jest.requireMock("src/features/skillsRanking/utils/createMessages");
    shouldSkipMarketDisclosure.mockReturnValue(true);
    const givenSessionId = 77777;
    const mockUpdate = jest.fn().mockResolvedValue({} as SkillsRankingState);
    jest.spyOn(SkillsRankingService, "getInstance").mockReturnValue({
      updateSkillsRankingState: mockUpdate,
      getConfig: () => ({
        config: {
          compensationAmount: "$1",
          jobPlatformUrl: "x",
          shortTypingDurationMs: 1,
          defaultTypingDurationMs: 1,
          longTypingDurationMs: 1,
        },
      }),
    } as unknown as SkillsRankingService);
    jest
      .spyOn(UserPreferencesStateService, "getInstance")
      .mockReturnValue({ getActiveSessionId: () => givenSessionId } as any);
    const actualOnFinish = jest.fn().mockResolvedValue(undefined);
    const givenState = createState(SkillsRankingPhase.RETYPED_RANK, 23);

    // WHEN render (component returns empty but effect runs)
    render(<SkillsRankingRetypedRank onFinish={actualOnFinish} skillsRankingState={givenState} />);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    await flush();

    // THEN service called and onFinish invoked
    // value remains 0 in RETYPED_RANK (state is not synced from props)
    expect(mockUpdate).toHaveBeenCalledWith(givenSessionId, SkillsRankingPhase.COMPLETED, 0);
    expect(actualOnFinish).toHaveBeenCalled();
  });

  test("should not proceed when no active session id", async () => {
    jest.spyOn(SkillsRankingService, "getInstance").mockReturnValue({
      updateSkillsRankingState: jest.fn(),
      getConfig: () => ({
        config: {
          compensationAmount: "$1",
          jobPlatformUrl: "x",
          shortTypingDurationMs: 1,
          defaultTypingDurationMs: 1,
          longTypingDurationMs: 1,
        },
      }),
    } as unknown as SkillsRankingService);
    jest.spyOn(UserPreferencesStateService, "getInstance").mockReturnValue({ getActiveSessionId: () => null } as any);
    const actualOnFinish = jest.fn();
    const givenState = createState(SkillsRankingPhase.RETYPED_RANK);

    render(<SkillsRankingRetypedRank onFinish={actualOnFinish} skillsRankingState={givenState} />);
    setSliderValue(10);
    const givenSubmitButton = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_RETYPED_RANK_SUBMIT_BUTTON);
    await waitFor(() => expect(givenSubmitButton).not.toBeDisabled());
    act(() => {
      givenSubmitButton.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    expect(console.error).toHaveBeenCalled();
    expect(actualOnFinish).not.toHaveBeenCalled();
  });

  test("should ignore submit in replay mode", async () => {
    const mockUpdate = jest.fn();
    jest.spyOn(SkillsRankingService, "getInstance").mockReturnValue({
      updateSkillsRankingState: mockUpdate,
      getConfig: () => ({
        config: {
          compensationAmount: "$1",
          jobPlatformUrl: "x",
          shortTypingDurationMs: 1,
          defaultTypingDurationMs: 1,
          longTypingDurationMs: 1,
        },
      }),
    } as unknown as SkillsRankingService);
    jest.spyOn(UserPreferencesStateService, "getInstance").mockReturnValue({ getActiveSessionId: () => 1 } as any);
    const givenState = createState(SkillsRankingPhase.COMPLETED);

    render(<SkillsRankingRetypedRank onFinish={jest.fn()} skillsRankingState={givenState} />);
    const givenSubmitButton = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_RETYPED_RANK_SUBMIT_BUTTON);
    expect(givenSubmitButton).toBeDisabled();
    act(() => {
      givenSubmitButton.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    expect(mockUpdate).not.toHaveBeenCalled();
    expect(console.error).not.toHaveBeenCalled();
  });

  test("snapshot: initial render in RETYPED_RANK (non-replay)", () => {
    // GIVEN default state in RETYPED_RANK
    jest.spyOn(SkillsRankingService, "getInstance").mockReturnValue({
      updateSkillsRankingState: jest.fn(),
      getConfig: () => ({
        config: {
          compensationAmount: "$1",
          jobPlatformUrl: "x",
          shortTypingDurationMs: 1,
          defaultTypingDurationMs: 1,
          longTypingDurationMs: 1,
        },
      }),
    } as unknown as SkillsRankingService);
    jest.spyOn(UserPreferencesStateService, "getInstance").mockReturnValue({ getActiveSessionId: () => 1 } as any);
    const givenState = createState(SkillsRankingPhase.RETYPED_RANK);

    // WHEN render
    const { container } = render(<SkillsRankingRetypedRank onFinish={jest.fn()} skillsRankingState={givenState} />);

    // THEN expect snapshot to match
    expect(container).toMatchSnapshot();
  });
});
