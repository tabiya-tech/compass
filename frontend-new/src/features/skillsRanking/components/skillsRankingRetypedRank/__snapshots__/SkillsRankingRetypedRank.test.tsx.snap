// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillsRankingRetypedRank snapshot: initial render in RETYPED_RANK (non-replay) 1`] = `
<div>
  <div
    class="MuiBox-root css-1s6oyup"
    data-testid="skills-ranking-retyped-rank-container-eb90de4c-2462-4b6d-8b9c-1b5c6ae64129"
    origin="COMPASS"
  >
    <div
      class="MuiBox-root css-8atqhb"
    >
      <div
        class="MuiBox-root css-u9ttk8"
        data-testid="chat-message-bubble-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
        origin="COMPASS"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-13zm2uq-MuiTypography-root"
          data-testid="chat-message-bubble-message-text-6e685eeb-2b54-432a-8b66-8a81633b3981"
        >
          As a last question, let's recall what I told you further above just regarding your own fit with opportunities:
           
          <strong>
            check again what I said three messages ago, how many percent of opportunities
          </strong>
           on
           
          x
           do you meet the key skills for?
        </p>
        <div
          class="MuiBox-root css-0"
          data-testid="chat-message-bubble-message-footer-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
        >
          <div
            class="MuiBox-root css-19midj6"
          >
            <input
              aria-label="Retyped rank percentile slider"
              data-testid="skills-ranking-retyped-rank-slider-eb90de4c-2462-4b6d-8b9c-1b5c6ae64129"
              type="range"
              value="0"
            />
            <div
              class="MuiBox-root css-k6jh9e"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation css-fph7a5-MuiButtonBase-root-MuiButton-root"
                data-testid="skills-ranking-retyped-rank-submit-button-eb90de4c-2462-4b6d-8b9c-1b5c6ae64129"
                disabled=""
                tabindex="-1"
                type="button"
              >
                Submit
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-1qcd69q"
        data-testid="chat-message-footer-layout-container-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
      >
        <span
          class="MuiTypography-root MuiTypography-caption css-1l3yql3-MuiTypography-root"
          data-testid="chat-message-timestamp-c9253326-05ee-4bff-9d43-7852ca78a033"
        >
          sent 
          just now
        </span>
      </div>
    </div>
  </div>
</div>
`;
